default/scripting/buildings/TERRAFORM.focs.txt

```diff
@@ -87,36 +87,33 @@ TARGET_TERRAFORMING
                     * Target.HabitableSize
                       // calculating multiple steps: 1+2+..+N == (1+N)*N/2
-                    * max(1, (1 + (PlanetTypeDifference from = Target.OriginalType to = Target.NextBestPlanetType))
-                             * (PlanetTypeDifference from = Target.OriginalType to = Target.NextBestPlanetType) / 2 )
+                    * max(1, (1 + (PlanetTypeDifference from = Target.PlanetType to = Target.NextBestPlanetType))
+                             * (PlanetTypeDifference from = Target.PlanetType to = Target.NextBestPlanetType) / 2 )
+                      // Giving a discount to amend opportunity cost by not multi-stepping
+                      // per extra step 10%, fast linear formula
+                    * (1 - 0.1 * ((PlanetTypeDifference from = Target.PlanetType to = Target.NextBestPlanetType) - 1))
                     * [[BUILDING_COST_MULTIPLIER]]
                     * (1 + (2 * Statistic If condition = And [Source EmpireHasAdoptedPolicy empire = Source.Owner name = "PLC_ENVIRONMENTALISM"]))
-    buildtime =  (8 - (4 * Statistic If condition = And [Source EmpireHasAdoptedPolicy empire = Source.Owner name = "PLC_TERRAFORMING"]))
-               * PlanetTypeDifference from = Target.OriginalType to = Target.NextBestPlanetType
+    buildtime =  (PlanetTypeDifference from = Target.PlanetType to = Target.NextBestPlanetType)
+                    * (100 - 10 * ((PlanetTypeDifference from = Target.PlanetType to = Target.NextBestPlanetType) - 1)) / 100
+                    * (12 - (6 * Statistic If condition = And [Source EmpireHasAdoptedPolicy empire = Source.Owner name = "PLC_TERRAFORMING"]))
     location = And [
         Planet
         Not Contains Building name = "BLD_TERRAFORM"
-[[FOR_ALL_TERRAFORM_PLANET_TYPES(        Not Contains Building name = "BLD_TERRAFORM_,")]]
-[[FOR_ALL_TERRAFORM_PLANET_TYPES(        Not Enqueued type = Building name = "BLD_TERRAFORM_,")]]
         OwnedBy empire = Source.Owner
         TargetPopulation low = 1
         Not Planet type = [Asteroids GasGiant]
         Not Planet environment = [Uninhabitable Good]
         Not Contains Building name = "BLD_GAIA_TRANS"
         Not HasSpecial name = "GAIA_SPECIAL"
+[[FOR_ALL_TERRAFORM_PLANET_TYPES(        Not Contains Building name = "BLD_TERRAFORM_,")]]
+[[FOR_ALL_TERRAFORM_PLANET_TYPES(        Not Enqueued type = Building name = "BLD_TERRAFORM_,")]]
         // Hide building if the species can not improve (Check species' highest available target environment tag)
         Not HasTag name = "NO_TERRAFORM"
-        Not And [
-            HasTag name = "TERRAFORM_TO_POOR"
-            Or [ Planet environment = Adequate Planet environment = Poor ]
-        ]
-        Not And [
-            HasTag name = "TERRAFORM_TO_ADEQUATE"
-            Or [ Planet environment = Adequate ]
-        ]
+        (0 < PlanetTypeDifference from = LocalCandidate.PlanetType to = LocalCandidate.NextBestPlanetType)
```

Comment:
Thus can probably just be an == test between the next best and current types

Reply:
Parser sais no after the following change
```
-        (0 < PlanetTypeDifference from = LocalCandidate.PlanetType to = LocalCandidate.NextBestPlanetType)
+        (LocalCandidate.PlanetType == LocalCandidate.NextBestPlanetType)
```
```
default/scripting/buildings/TERRAFORM.focs.txt:885:8: Parse error.  Expected ] here:
        Not Enqueued type = Building name = "BLD_TERRAFORM_INFERNO"
        Not Enqueued type = Building name = "BLD_TERRAFORM_RADIATED"
        Not Enqueued type = Building name = "BLD_TERRAFORM_BARREN"
        // Hide building if the species can not improve (Check species' highest available target environment tag)
        Not HasTag name = "NO_TERRAFORM"
        (LocalCandidate.PlanetType == LocalCandidate.NextBestPlanetType)
        ^
```

Reply:
if we had arithmetics for enums we could also kick out PlanetTypeDifference everywhere and write instead
`abs(LocalCandidate.PlanetType - LocalCandidate.NextBestPlanetType))`

Reply:
so for this PR i wontfix this

Reply:
Try with latest master? It should work for those PlanetType valuerefs...

Reply:
syntax check OK
```
(LocalCandidate.PlanetType == LocalCandidate.NextBestPlanetType)
```

syntax check FAIL (Parse error.  Expected real number expression here):
```
((LocalCandidate.PlanetType) == (LocalCandidate.NextBestPlanetType))
  ^
```

Reply:
Do you particularly like adding extra `()` around sub-expressions without ambiguous operator precedence issues?

Reply:
Yes, this works. Though I am a little uneasy of the class of new bugs being possible (e.g. comparing values of different enum types).

Reply:
So far only PlanetType enums are supported by the parser.

Reply:
> Do you particularly like adding extra `()` around sub-expressions without ambiguous operator precedence issues?

yes, thats my personal hobby. 

no, those were left over from a previous more complex calc; i really don't care about those, just wanted to provide feedback to you. also note i did not open an issue.

and in this case it might even be helpful that it does not work, as doing a complicated calculation using enums is probably the wrong thing to do. it will of course only show up sporadically (when it is used in an integer expression)